File: lib/ES/ES/ES.rb
#! /usr/bin/env

require_relative 'c_es'

=begin rdoc
El modulo ES engloba a las clases que pertenecen
al Enfriamiento Simulado
=end
module ES

	# La clase ES trata de representar los conceptos de Enfriammiento Simulado
	class ES
    # El atributo coste_solucion_actual almacena el coste de la solucion actual
    # de manera que se puedan hacer los calculos de enfriamiento simulado 
    # de manera rapida y eficiente
    attr_accessor :coste_solucion_actual

    # El atributo aceptacion devuelve la probabilidad en tanto por uno de
    # aceptar una solucion diferente a la actual
    attr_reader :aceptacion

		# Constructor de la clase, recibe como parametros:
		#	- valor_inicio: Indica el valor de inicio de la temperatura, debe de ser
  	#		numerico y solo puede oscilar entre 0-1 (ambos inclusive)
		#	- tipo: Indica el tipo de función que se usara para disminuir la temperatura
  	#		Actualmente solo puede ser :geometrica
		#	- coeficiente: Indica el valor del coeficiente usado en la función de disminución
  	#		de la temperatura
    # - aceptacion: probabilidad de aceptacion en tanto por uno de una solucion distinta
  		def initialize(valor_inicio: 1, tipo: :geometrica, coeficiente: 0.85, aceptacion: 1)
  			raise TypeError, "valor_inicio debe de ser numerico" unless valor_inicio.kind_of? Numeric
  			raise TypeError, "tipo de valer :geometrica" unless tipo.eql? :geometrica
  			raise TypeError, "coeficiente debe de ser un valor numerico" unless coeficiente.kind_of? Numeric
        raise TypeError, "aceptacion debe de ser un valor numerico" unless aceptacion.kind_of? Numeric

  			if coeficiente >= 1 or coeficiente <= 0
  				raise TypeError, "coeficiente debe de ser menor o igual que 1 (ambos exclusive)"
  			end

  			@valor_inicio = valor_inicio
  			@tipo = tipo
  			@coeficiente = coeficiente
  			@temperatura = valor_inicio
        @aceptacion = aceptacion
        @coste_solucion_actual = 1
  		end
	end
end

File: lib/ES/ES/ES.c
#include "ES.h"
#include <stdlib.h>

/*
Este método devuelve la temperatura actual del algoritmo
*/
VALUE method_temperatura(VALUE self)
{
	return rb_iv_get(self, "@temperatura");
}

/*
Este método devuelve un valor booleano True o False de forma aleatoria.
La probabilidad de que devuelva True sera mayor contra más alta sea
la temperatura.
*/
VALUE method_probabilidad(VALUE self)
{
	//ID method_rand = rb_intern("rand");
	VALUE valorInicio = rb_iv_get(self, "@temperatura_inicio");
	VALUE valorTemperatura = rb_iv_get(self, "@temperatura");
	//VALUE valorAleatorio = rb_funcall(rb_cObject, method_rand, 1, valorInicio);
	double valorAleatorio = rb_genrand_real();
	
	if(valorAleatorio <= NUM2DBL(valorTemperatura) / NUM2DBL(valorInicio))
	{
		return Qtrue;
	}
	else
	{
		return Qfalse;
	}
}

/*
Disminuye la temperatura interna de la clase de manera que haya menor
probabilidad de conseguir un valor True.

La función de disminución de la temperatura viene determinada por el
tipo y por el coeficiente de cambio determinado en la clase
*/
VALUE method_enfriar(VALUE self)
{
	VALUE coeficiente;
	VALUE temperaturaActual;

	coeficiente = rb_iv_get(self, "@coeficiente");
	temperaturaActual = rb_iv_get(self, "@temperatura");

	temperaturaActual = DBL2NUM(NUM2DBL(temperaturaActual) * NUM2DBL(coeficiente));
	rb_iv_set(self, "@temperatura", temperaturaActual);
	
	return temperaturaActual;
}

/*
Devuelve el coeficiente de cambio usado en la función de temperatura
*/
VALUE method_coeficiente(VALUE self)
{
	return rb_iv_get(self, "@coeficiente");
}

/*
Restaura la temperatura a su punto inicial
*/
VALUE method_reset(VALUE self)
{
	VALUE valorInicial = rb_iv_get(self, "@temperatura_inicio");
	rb_iv_set(self, "@temperatura", valorInicial);
	return valorInicial;
}

/*
Constructor de la clase ES. Recibe como parámetros:
- temperatura: El valor de la temperatura inicial. Este valor debe de ser un numero positivo
- coeficiente: Valor del coeficiente. Este valor debe de ser un numero entre 0-1 exclusive
*/
VALUE method_es_initialize(VALUE self, VALUE temperatura, VALUE coeficiente)
{
	if((TYPE(temperatura) != T_FIXNUM) && (TYPE(temperatura) != T_FLOAT))
	{
		rb_raise(rb_eTypeError, "La temperatura de inicio debe de ser un numero\n");
	}

	if((TYPE(coeficiente) != T_FIXNUM) && (TYPE(coeficiente) != T_FLOAT))
	{
		rb_raise(rb_eTypeError, "El coeficiente debe de ser un flotante\n");
	}

	if(NUM2DBL(temperatura) < 1)
	{
		rb_raise(rb_eTypeError, "La temperatura no puede ser negativa\n");
	}

	if((NUM2DBL(coeficiente) <= 0) || (NUM2DBL(coeficiente) >= 1))
	{
		rb_raise(rb_eTypeError, "El coeficiente debe oscilar entre 0-1 (exclusive)\n");
	}

	rb_iv_set(self, "@temperatura", temperatura);
	rb_iv_set(self, "@coeficiente", coeficiente);
	rb_iv_set(self, "@temperatura_inicio", temperatura);
	return self;
}


void Init_ES()
{
	module_es = rb_define_module("ES");
	class_es = rb_define_class_under(module_es, "ES", rb_cObject);
	rb_define_method(class_es, "initialize", method_es_initialize, 2);
	rb_define_method(class_es, "temperatura", method_temperatura, 0);
	rb_define_method(class_es, "probabilidad", method_probabilidad, 0);
	rb_define_method(class_es, "enfriar", method_enfriar, 0);
	rb_define_method(class_es, "coeficiente", method_coeficiente, 0);
	rb_define_method(class_es, "reset", method_reset, 0);
}
